### PR TITLE
objdetect: clear stale outputs in CharucoDetector::detectDiamonds

### PR DESCRIPTION
**Pull request by jeevan6996**
## Summary
- clear `diamondCorners`/`diamondIds` outputs at the beginning of `CharucoDetector::detectDiamonds`
- prevent stale detections from previous calls when the current call has fewer than 4 markers or finds no diamonds
- add a regression test that pre-fills outputs, calls `detectDiamonds` with 3 markers, and verifies outputs are empty

Fixes #28783.

## Testing
- built `opencv_test_objdetect` locally
- ran `opencv_test_objdetect --gtest_filter=Charuco.detectDiamondsClearsOutputsWithLessThanFourMarkers`
- result: PASS

### CODE REVIEW

File: modules/objdetect/src/aruco/charuco_detector.cpp [hunk 1]
```diff
@@ -400,6 +400,9 @@ void CharucoDetector::detectDiamonds(InputArray image, OutputArrayOfArrays _diam
     CV_Assert(getBoard().getChessboardSize() == Size(3, 3));
     CV_Assert((inMarkerCorners.empty() && inMarkerIds.empty() && !image.empty()) || (inMarkerCorners.total() == inMarkerIds.total()));
 
+    if (_diamondCorners.needed()) _diamondCorners.release();
```
Comment (asmorkalov) [inline modules/objdetect/src/aruco/charuco_detector.cpp:403]: I propose to move it to line 530 like
```
if(diamondIds.size() > 0ull) 
{
...
}
else
{
// release here
}

**Review by asmorkalov — APPROVED** [review]
